File: solutions/lua/run-length-encoding/1/run-length-encoding.lua
return {
  encode = function(s)
    if #s == 1 then
      error("")
    end
    local result = ""
    local i = 1
    while i <= #s do
      local tmp = s:sub(i, i)
      local count = 0
      while tmp == s:sub(i, i) do
        count = count + 1
        i = i + 1
      end
      if count == 1 then
        result = result .. tmp
      else
        result = result .. count .. tmp
      end
    end
    return result
  end,

  decode = function(s)
    local result = ""
    local num = ""
    local char = ""
    local i = 1
    while i <= #s do
      local char = s:sub(i, i)
      if tonumber(char) ~= nil then
        num = num .. char
        goto continue
      end
      num = (num or "")
      if num ~= "" then
        result = result .. string.rep(char, tonumber(num))
        num = ""
      else
        result = result .. char
      end
      ::continue::
      i = i + 1
    end
    return result
  end
}

File: solutions/go/run-length-encoding/1/run_length_encoding.go
package encode

import (
	"strconv"
	"strings"
	"unicode"
)

type rle_s struct {
	ch    byte
	count int
}

func RunLengthEncode(input string) string {
	test := []rle_s{}

	var prev rune
	for _, ii := range input {
		if prev == ii {
			test[len(test)-1].count++
		} else {
			test = append(test, rle_s{byte(ii), 1})
		}
		prev = ii
	}

	result := []byte{}
	for _, ii := range test {
		if ii.count == 1 {
			result = append(result, ii.ch)
		} else {
			var01 := []byte(strconv.Itoa(ii.count))
			result = append(result, var01...)
			result = append(result, ii.ch)
		}
	}
	return string(result)
}

func RunLengthDecode(input string) string {
	number := []byte{}
	result := ""
	for _, ii := range input {
		if unicode.IsNumber(ii) {
			number = append(number, byte(ii))
			continue
		} else {
			ch := string(ii)
			str := ""
			if num, ok := strconv.Atoi(string(number)); ok == nil {
				str = strings.Repeat(ch, num)
				number = []byte{}
			} else {
				str = ch
			}
			result += str
		}
	}
	return result
}
